### ISSUE TITLE
KeyedVectors TODOs

### ISSUE THREAD

**Issue by darvid7** [RaRe-Technologies/gensim#2166]
#### Description
Hi! I'm new here :) I was digging through the `Doc2Vec` code and came across a bunch of TODOs in `KeyedVectors`.

 I was wondering if TODOs such as https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/keyedvectors.py#L1601 in [KeyedVectors](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/keyedvectors.py) can be implemented using `model.infer()` like in https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/keyedvectors.py#L1801

If so I'll be happy to submit a PR.

#### Steps/Code/Corpus to Reproduce

*N/A*

#### Expected Results

*N/A*

#### Actual Results

*N/A*

#### Versions

Darwin-15.6.0-x86_64-i386-64bit
Python 3.6.4 (default, Jan  6 2018, 11:49:38) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
NumPy 1.15.1
SciPy 1.1.0
gensim 3.5.0
FAST_VERSION 0

**Comment by piskvorky** [comment]
Thanks for the offer! I believe that TODO refers to accepting "raw vectors" (np.ndarray), in addition to the str/int vector labels:

```python
sims = model.most_similar(positive=[np.array([1, 2, 3]), np.array([4, 5, 6]), 'doctag', …)
```

The idea is that people can pass in their own vectors, no matter where they came from, such as from `infer()`:

```python
sims = model.most_similar(positive=[model.infer_vector(words), np.array([4, 5, 6]), 'doctag', …)
```

(that is, not doing `infer` internally, but being able to accept its output)

CC @gojomo @menshikh-iv agree?

**Comment by gojomo** [comment]
AFAIK that method already accepts raw vectors. I'm unsure of the intent of the TODO. (It might have meant take the raw documents – as lists-of-tokens or `TaggedDocument` instances – directly. But I think that'd be unnecessary further parameter overloading.)

**Comment by piskvorky** [comment]
If that's the case, then at least the documentation is wrong.

**Comment by menshikh-iv** [comment]
In this case, need to
- remove TODO
- fix parameter types (docstring)

**Comment by rsdel2007** [comment]
@menshikh-iv, is this issue solved? Can I work on this?

**Comment by rsdel2007** [comment]
@menshikh-iv So, I have to implement TODO except for `most_similar`?

**Comment by menshikh-iv** [comment]
@rsdel2007 almost yes, read again https://github.com/RaRe-Technologies/gensim/issues/2166#issuecomment-416101174

**Comment by rsdel2007** [comment]
@menshikh-iv , How to implement "accept vector"? Please guide.

**Comment by menshikh-iv** [comment]
@rsdel2007 in same way as this already done in https://github.com/RaRe-Technologies/gensim/blob/b452a5b59f2f474dbbd275d0838c45df4d3c5aac/gensim/models/keyedvectors.py#L486

check out ndarray related parts in this function (word can be a vector, not only str)

**Comment by Naba7** [comment]
is this issue closed or open?

**Comment by PredictedLife** [comment]
The documentation does not yet describe that you can also give inferred vectors to the method most_similar.

If you call most_similar method with an inferred vector which is also present in the training set a top-n list of most similar documents including consine similarities is returned.

If the document was also present in the training set the closest returned most_similar document is the same document. Suprisingly the cosine_similarity is 0.86 but never 1.

Is this a bug? Or is there any explanation for this behaviour?
See issue: https://github.com/RaRe-Technologies/gensim/issues/2915

**Comment by gojomo** [comment]
Yes, the `most_similar()` doc-comment should be clearer about the option to supply raw vectors, not just keys, as part of its `positive` or `negative` parameters. (There's a [slight improvement in development code](https://github.com/RaRe-Technologies/gensim/blob/b30888342e454a58ea000edc8b985a0b7ac5451b/gensim/models/keyedvectors.py#L613), where `ndarray` is listed as an allowed type in the parameter specs, but the descriptive text is still incomplete.)

When you re-infer a vector for the same list-of-words as appeared in training, you won't and shouldn't expect to get the exact same vector - for reasons covered in Q11 & Q 12 of the Gensim FAQ: https://github.com/RaRe-Technologies/gensim/wiki/recipes-&-faq#q11-ive-trained-my-word2vecdoc2vecetc-model-repeatedly-using-the-exact-same-text-corpus-but-the-vectors-are-different-each-time-is-there-a-bug-or-have-i-made-a-mistake-2vec-training-non-determinism
